### ISSUE TITLE
Documentation not showing up through rust-analyzer when include_str! is used

### ISSUE THREAD

**Issue by udoprog** [tokio-rs/axum#929]
The way [documentation is defined for certain functions](https://docs.rs/axum/latest/src/axum/routing/mod.rs.html#407) prevents it from showing up through rust-analyzer (see screenshot).

This is what you see when you use `into_make_service()`:
![image](https://user-images.githubusercontent.com/111092/162616566-99e19cf1-5b9e-4b18-aa9f-7c99a6eee8d0.png)

This is what you see when you use `into_make_service_with_connect_info()`:
![image](https://user-images.githubusercontent.com/111092/162616587-2f2b55cd-011b-41e3-b4e6-a0e3de236005.png)

I don't know if this should count as a rust-analyzer issue, but it could be worth considering to improve usability of APIs with the current state of tooling.

**Comment by davidpdrsn** [comment]
Hm thats kinda annoying 😕 Its definitely a rust-analyzer bug. Using `#[doc = include_str!(...)]` was highlighted as a use case [in the announcement of Rust 1.54](https://blog.rust-lang.org/2021/07/29/Rust-1.54.0.html#attributes-can-invoke-function-like-macros). I'd be sad having to remove it. On the other hand I understand the frustration. Docs are read more times than they're written.

Would you mind filing an issue with rust-analyzer?

**Comment by udoprog** [comment]
Just looked, here are the related issues:
* rust-analyzer/rust-analyzer#11137
* rust-analyzer/rust-analyzer#8092

**Comment by davidpdrsn** [comment]
I'll close this for now. I think its up to RA to fix this and at least seems they're aware of the issue.

**Comment by joshka** [comment]
It's true that this is a shortcoming of RA which hasn't been fixed in 3 years likely because it's difficult to fix. But it's also a problem that appears when navigating to the source in an IDE. It's pretty useful to actually see the documentation inline with the code that it documents and I'd suggest that it would be reasonable to want this. I find it really really annoying to hit the dead end when viewing axum's source code, whether that via github, or in my IDE.

I'm going to submit a PR which fixes this, and would suggest re-opening this issue to discuss the finer points of whether this should happen.

**Comment by jplatte** [comment]
I actually like when longer docs are their own file, when looking at the source code. I find the separate markdown file is generally better highlighted, not mostly green or whatever color doc comments are, plus the source file is less bloated. In VSCode with RA, I can Ctrl+click to jump to a file from `include!` or `include_str!`.

**Comment by joshka** [comment]
>I actually like when longer docs are their own file, when looking at the source code. I find the separate markdown file is generally better highlighted, not mostly green or whatever color doc comments are, plus the source file is less bloated. 

I can see how you might come to those viewpoints as someone intimately familiar with the source, but you're likely not the target audience of the docs. I wonder if you can you see that it's possible that users of Axum to come to the exact opposite viewpoint as you here?
- The color of comments is a personal choice that is just a theme setting away (I like the green personally)
- I don't see any meaningful difference in the highlighting of code in doc comments vs markdown files.
- While there are less lines per file, this realistically means that the source is more fragmented and less navigable. I'm forced to look at two places rather than one for the information I need. This is particularly meaningful when there's complexity in the types.

I didn't raise this issue after a single amount of frustration with this, but a few incidences of this frustration growing over time. It's a small papercut, and I'm definitely someone who is vocal about fixing small issues like this

> In VSCode with RA, I can Ctrl+click to jump to a file from include! or include_str!.

This doesn't seem like it's available anywhere I've tried when clicking these files. Is there a setting to enable this? (not that I could find)

Does it work when hovering over source code to get inline information about the nuance of a method call? (no)

Does it work when I've navigated to a source file from another project? (no)

Do intra-doc links work from the markdown files? (no)

Does it work when navigating the source in GitHub or other tooling? (no)

Is this a huge deal? Probably not. The workaround is just remembering all the time that a random selection of Axum functions, modules, etc. don't have docs that work in an IDE and so I always have to navigate to docs.rs (or local generated docs) instead of doing the usual thing. It's an annoying papercut though, which I hit almost every time I deal with Axum code, and it's easy to fix with (IMO) little downside and a decent upside.

**Comment by jplatte** [comment]
I'll reopen since I think you have some good points, but I'm still not sure whether we are going to change things.

Maybe some of the other maintainers want to weigh in as well. @mladedav @yanns @Turbo87 @GlenDC @tottoto

**Comment by udoprog** [comment]
I'm still inconvenienced by this. Every time I spin up a new project `Router` is unusable without documentation next to you. It's the only part about axum that bothers me consistently.

**Comment by yanns** [comment]
My 2 cents is that we should optimize for the developer experience.
So I'd vote to avoid the `include_str!` for now as not all the developer tooling is able to deal with that.

**Comment by joshka** [comment]
I case I wasn't clear on this, I'm also happy to submit a PR to fix this, but it takes a bit of effort to do it correctly. ("Replace include_str with text" would be a nice refactoring for rust-analyzer to add someday - https://github.com/rust-lang/rust-analyzer/issues/20494)

**Comment by Turbo87** [comment]
FWIW if we use `include_str!()` to e.g. reuse a Readme file as the main module docs, then I think this is a legitimate use case. if we only use it to avoid having the markdown inline in a doc comment but there is no reuse happening, then IMHO it's probably not worth the tradeoffs. from an initial look at the codebase, it seems like the latter is the case here, so I'd vote for inlining the `include_str!()` usages.

**Comment by offline-pixel** [comment]
Let me see

**Comment by jplatte** [comment]
Should be fixed with Monday's RA release: https://github.com/rust-lang/rust-analyzer/pull/21928

**Comment by joshka** [comment]
It does not appear to have fixed the problem:

Installed version:
<img width="300" height="217" alt="Image" src="https://github.com/user-attachments/assets/80080dba-5b20-45cd-b11a-b686da73ee71" />

https://github.com/tokio-rs/axum/blob/441216428893d13544f12722b54dcaaadd47135a/axum/src/error_handling/mod.rs#L1
Module level doc:
<img width="450" height="217" alt="Image" src="https://github.com/user-attachments/assets/f5315d2b-d8a3-49c9-a034-91ac66e94d9b" />

Function level doc:
https://github.com/tokio-rs/axum/blob/441216428893d13544f12722b54dcaaadd47135a/axum/src/routing/method_routing.rs#L1002
<img width="1008" height="560" alt="Image" src="https://github.com/user-attachments/assets/9e35710b-2f63-4b12-a577-86fad8b9ec33" />

(posted in the PR, just noting that this is not yet fixed)

**Comment by so1ve** [comment]
This should be fixed with the latest release of RA

**Comment by mladedav** [comment]
Yeah, I happened to check today and it did work for me, both in this repo itself and in examples which are a different workspace, though still dependency by path.

Thanks for the fix.
